### ISSUE TITLE
rook-operator should support custom ceph.conf settings

### ISSUE THREAD

**Issue by bzub** [rook/rook#571]
With standalone rook it's easy: pass `--ceph-config-override=ceph.conf`

### Implementation ideas
- The user supplies the name of a ConfigMap in the Cluster TPR
  - rook-operator inserts this into OSD/Mon/etc k8s definitions (mounts it as a file, points to it in CLI args)
  - User is responsible for creating the ConfigMap
- The user supplies a literal string in the Cluster TPR
  - rook-operator creates a ConfigMap and configures pods to use them
- The user supplies objects containing key/values in the Cluster TPR named after the ceph.conf sections they fall under
  - rook-operator translates these into ceph.conf INI format, creates ConfigMap, configures pods

The last idea is similar to how many Helm Charts work, and might look like this:
```yaml
configOverride:
  - global:
    osd_crush_update_on_start: false
    "osd pool default size": 2
```
Resulting in a config like:
```dosini
[global]
osd crush update on start = false
osd pool default size     = 2
```

### Toolbox
I'm thinking the toolbox doesn't need to apply custom ceph.conf settings.  I might be mistaken but I don't think any of those options affect clients.

**Comment by travisn** [comment]
great ideas @bzub. I would lean towards the first approach. The operator could always generate a configmap for each service: one for the monitors, one for osds, one for mds, etc. The configmap would be mapped as a file in the pods as you suggested. By default it is empty, then Rook users could update the configmap with their desired config. Though this would likely require a restart of a pod to pick up the settings.

**Comment by bzub** [comment]
I like that idea.  Start off with blank ConfigMap(s).  In that case I suppose the name(s) wouldn't even be an option, which I think is less prone to error.

**Comment by bzub** [comment]
We should test to see what an invalid config does to a cluster.  Warning message or crash-looping?

**Comment by jbw976** [comment]
Help me understand a bit better how k8s config maps getting mapped to pods as a file works.  does it get auto formatted into the expected ini format that ceph expects?

**Comment by bzub** [comment]
ConfigMaps can contain any number of "key: value" pairs, but they can only be consumed as literal strings I believe.  So no conversion unless a program in a pod does that conversion.  Most common use is to put the contents of a text file, for example a ceph.conf, in the ConfigMap as a string literal in the value portion, and "ceph.conf" as the key.

The values can also be used as environment variables or CLI arguments in a pod.

**Comment by bzub** [comment]
FYI there are fancy tools for generating ConfigMaps and Secrets with variables via templates.  Like [konfd](https://github.com/kelseyhightower/konfd).

**Comment by travisn** [comment]
Fixed by #674
